sled-agent/src/bootstrap/rss_handle.rs

```diff
@@ -0,0 +1,232 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! sled-agent's handle to the Rack Setup Service it spawns
+
+use super::client as bootstrap_agent_client;
+use super::discovery::PeerMonitorObserver;
+use super::params::SledAgentRequest;
+use crate::rack_setup::config::SetupServiceConfig;
+use crate::rack_setup::service::Service;
+use futures::stream::FuturesUnordered;
+use futures::Future;
+use futures::StreamExt;
+use omicron_common::backoff::internal_service_policy;
+use omicron_common::backoff::retry_notify;
+use omicron_common::backoff::BackoffError;
+use slog::Logger;
+use std::net::SocketAddrV6;
+use thiserror::Error;
+use tokio::sync::mpsc;
+use tokio::sync::oneshot;
+use tokio::task::JoinHandle;
+
+pub(super) struct RssHandle {
+    _rss: Service,
+    task: JoinHandle<()>,
+}
+
+impl Drop for RssHandle {
+    fn drop(&mut self) {
+        // NOTE: Ideally, with async drop, we'd await completion of the our task
+        // handler.
+        //
+        // Without that option, we instead opt to simply cancel the task to
+        // ensure it does not remain alive beyond the handle itself.
+        self.task.abort();
+    }
+}
+
+impl RssHandle {
+    // Start the Rack setup service.
+    pub(super) fn start_rss(
+        log: &Logger,
+        config: SetupServiceConfig,
+        peer_monitor: PeerMonitorObserver,
+    ) -> Self {
+        let (tx, rx) = rss_sled_agent_channel();
+
+        let rss = Service::new(
+            log.new(o!("component" => "RSS")),
+            config,
+            peer_monitor,
+            tx,
+        );
+        let log = log.new(o!("component" => "BootstrapAgentRssHandler"));
+        let task = tokio::spawn(async move {
+            handle_rss_requests(&log, rx).await;
+        });
+        Self { _rss: rss, task }
+    }
+}
+
+async fn handle_rss_requests(log: &Logger, requests: RssSledAgentReceiver) {
+    requests
+        .initialize_sleds(log, |bootstrap_addr, request| async move {
```

Comment:
IDK if it's just me, but I'm finding this control flow a little complex. To be clear:

- `RssHandle` spawns a task which runs `handle_rss_requests
- This calls `RssSledAgentReceiver::initialize_sleds`, which acts on a `Fn`
- This reads from an mpsc where a `RssSledAgentSender::initialize_sleds` has been called
- In response, the `Fn` is invoked on each request
- These functions are converted to futures, we await them serially (by calling `futs.next()` in a while loop

Why bother with the `Fn`? There's only one implementation, and it's owned by this file. Why not have `handle_rss_requests` call `RssSledAgentReceiver::initialize_sleds`, and call this code in-line (which itself calls `initialize_sled_agent`, with some logging) rather than passing a `Fn`? I think it would make the flow slightly easier to follow.

Reply:
Mmm, this may be overly complex because it grew out of an initial implementation where I _was_ taking requests one at a time instead of a `Vec`. I'll take a shot at your suggested simplification; thanks.

Reply:
After inlining there was no reason to even keep `handle_rss_requests()` at all: 6425915b

Reply:
Nice, this is *way* easier for me to read!